File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        solve(board);
    }

    private boolean solve(char[][] board) {
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                if (board[row][col] != '.'){
                    continue;
                }
                for (char k = '1'; k <= '9'; k++) {
                    if (isValid(board, row, col, k)) {
                        board[row][col] = k;
                        if (solve(board)){
                            return true;
                        }
                        board[row][col] = '.';
                    }
                }
                return false;
            }
        }
        return true;
    }

    private boolean isValid(char[][] board, int row, int col, char k) {
        for(int i = 0; i < 9; i++) {
            if ( board[row][i] == k) {
                return false;
            }
        }

        for(int i = 0; i < 9; i++) {
            if ( board[i][col] == k) {
                return false;
            }
        }

        for (int i = (row / 3) * 3; i < (row / 3) * 3 + 3; i++) {
            for (int j = (col / 3) * 3; j < (col / 3) * 3 + 3; j++) {
                if (board[i][j] == k) {
                    return false;
                }
            }
        }

        return true;

    }
}

File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        
        n = len(board)

        def isValid(row: int, col: int, val: str) -> bool:
            # Check row
            for j in range(n):
                if board[row][j] == val:
                    return False

            # Check column
            for i in range(n):
                if board[i][col] == val:
                    return False

            # Check square
            for i in range(row // 3 * 3, row // 3 * 3 + 3):
                for j in range(col // 3 * 3, col // 3 * 3 + 3):
                    if board[i][j] == val:
                        return False

            return True

        def solve():
            for i in range(n):
                for j in range(n):
                    if board[i][j] == ".":
                        for val in range(1, 10):
                            if isValid(i, j, str(val)):
                                board[i][j] = str(val)
                                if solve():
                                    return True
                                board[i][j] = "."
                        return False
            return True
        
        solve()